security/advancedtls/advancedtls.go

```diff
@@ -54,17 +54,34 @@ type VerificationFuncParams struct {
 	Leaf *x509.Certificate
 }
 
-// VerificationResults contains the information about results of
-// CustomVerificationFunc.
-// VerificationResults is an empty struct for now. It may be extended in the
+// DEPRECATED: Renamed to `HandshakeVerificationInfo`
+// VerificationFuncParams contains parameters available to users when
+// implementing CustomVerificationFunc.
+// The fields in this struct are read-only.
+type VerificationFuncParams = HandshakeVerificationInfo
+
+// PostHandshakeVerificationResults contains the information about results of
+// PostHandshakeVerificationFunc.
+// PostHandshakeVerificationResults is an empty struct for now. It may be extended in the
 // future to include more information.
-type VerificationResults struct{}
+type PostHandshakeVerificationResults struct{}
+
+// DEPRECATED: Renamed to `PostHandshakeVerificationResults`
+type VerificationResults = PostHandshakeVerificationResults
+
+// PostHandshakeVerificationFunc is the function defined by users to perform
+// custom verification checks after chain building and regular handshake
+// verification has been completed.
+// PostHandshakeVerificationFunc returns nil
+// if the authorization fails; otherwise returns an empty struct.
```

Comment:
I don't understand this.

The convention is that functions reutrn `<zero value AKA nil here>, error` or error and `_, nil` on success.  If `nil, nil` is illegal, then that makes the API confusing and hard to use -- in which case you should have it return a value instead of a pointer.

Reply:
It returns `nil, error` in the case the authorization fails, `<currently empty struct>, nil` on success

We discussed and don't think there is harm in keeping the struct even though it's currently unused - there's a good chance it could be in the future and it's already implemented as such

Reply:
OK - it's a little unusual but should be OK.  If that's what you want you should make sure you're panicking if you call it and the function ever returns `nil, nil`.

Reply:
I don't think there's a compelling reason to return a pointer vs. a struct if that is what is making this unclear, but that would represent a breaking behavior change. Looking at it, I don't think `nil, nil` is illegal, it just means there's no information to pass back out.

I added some text to the comment to make it a little more clear.

That being said, looking at it again today with fresh eyes, it's called [here](https://github.com/grpc/grpc-go/blob/34de5cf4832ffa43dec384ddb20c91ca0805954f/security/advancedtls/advancedtls.go#L581-L582) in our code. And the struct is always thrown away. So I'm rethinking the value of keeping this in the API. Will discuss with my team and add another comment soon.

Reply:
It says "otherwise returns an empty struct" which implies that `nil, nil` is invalid.  If `nil, nil` is valid, then you can just delete this part of the documentation.  Returning `<zero>, error` or `_, nil` is just how regular Go conventions always work.

Reply:
Ah I see, I'll remove that part and then this is resolved? I think that it's more pain that its worth to remove the struct return

Reply:
PTAL